search/tasks.py

```diff
@@ -68,22 +68,32 @@ def index_program_enrolled_users(program_enrollment_ids):
     _index_program_enrolled_users(program_enrollments)
 
     # Send email for profiles that newly fit the search query for an automatic email
-    post_indexing_handler(program_enrollment_ids)
+    post_indexing_handler(program_enrollments)
 
 
 @app.task
-def index_users(user_ids):
+def index_users(user_ids, check_if_changed=False):
     """
     Index users' ProgramEnrollment documents
 
     Args:
         user_ids (list of int): Ids of users to update in the Elasticsearch index
+        check_if_changed (bool):
+            If true, read the document from elasticsearch before indexing and
+            check if the serialized value would be different.
     """
-    _index_users(user_ids)
+    enrollments = list(ProgramEnrollment.objects.filter(user__in=user_ids))
 
-    # Send email for profiles that newly fit the search query for an automatic email
-    enrollment_ids = list(ProgramEnrollment.objects.filter(user__in=user_ids).values_list("id", flat=True))
-    post_indexing_handler(enrollment_ids)
+    if check_if_changed:
+        enrollments = [
+            enrollment for enrollment in enrollments if _document_needs_updating(enrollment)
```

Comment:
This seems like it could result in additional overhead in cases where the enrollment/profile *was* updated because we end up querying for and serializing the data twice. Once in `_document_needs_updating` and the other in `_get_program_enrolled_users_documents` which is called by `index_program_enrolled_users`. I suspect you place the check here so we never skip documents when rebuilding the index, but I'm a little concerned we're trading off overhead on one system for overhead on another.

Does it make sense to carry `check_if_changed` all the way to `_get_program_enrolled_users_documents` and simply not yield there instead?

Reply:
I'm not sure I follow, where is the overhead tradeoff happening?

Reply:
We're hitting reindexing less, but for those things we do index, we're computing the document twice now (including all the DB queries to do so). At worst, we've doubled our overhead on the MM side.

Reply:
The information on whether or not to reindex needs to exist both in this function (to decide on the post indexing tasks) and in the indexing function. In order to avoid serializing twice we would need to do it here and pass down the dict into the other functions. Also, currently `document_needs_updating` will do a Elasticsearch request for each document, which would change indexing from a one request per chunk operation to one per document.

I think ideally we would have a stream of ES documents and `ProgramEnrollment`s sorted by id, do the comparison and reindex the modified ones but it would require some refactoring, and `index_users` currently only handles one user at a time so I'm not sure it's worth it.

Reply:
Yeah, I'm starting to see your point there. I agree with the current design this is probably the best approach even if not optimal.